controllers/reconcilers/kserve_raw_route_reconciler.go

```diff
@@ -0,0 +1,195 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strconv"
+)
+
+const (
+	isvcServicePort = 8888
+	authServicePort = 8443
+)
+
+var _ SubResourceReconciler = (*KserveRawRouteReconciler)(nil)
+
+type KserveRawRouteReconciler struct {
+	client         client.Client
+	routeHandler   resources.RouteHandler
+	deltaProcessor processors.DeltaProcessor
+}
+
+func NewKserveRawRouteReconciler(client client.Client) *KserveRawRouteReconciler {
+	return &KserveRawRouteReconciler{
+		client:         client,
+		routeHandler:   resources.NewRouteHandler(client),
+		deltaProcessor: processors.NewDeltaProcessor(),
+	}
+}
+
+func (r *KserveRawRouteReconciler) Reconcile(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error {
+	log.V(1).Info("Reconciling Generic Route for Kserve InferenceService")
+
+	// Create Desired resource
+	desiredResource, err := r.createDesiredResource(ctx, log, isvc)
+	if err != nil {
+		return err
+	}
+
+	// Get Existing resource
+	existingResource, err := r.getExistingResource(ctx, log, isvc)
+	if err != nil {
+		return err
+	}
+
+	// Process Delta
+	if err = r.processDelta(ctx, log, desiredResource, existingResource); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *KserveRawRouteReconciler) Delete(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error {
+	log.V(1).Info("Deleting Kserve inference service generic route")
+	return r.routeHandler.DeleteRoute(ctx, types.NamespacedName{Name: isvc.Name, Namespace: isvc.Namespace})
+}
+
+func (r *KserveRawRouteReconciler) Cleanup(_ context.Context, _ logr.Logger, _ string) error {
+	// NOOP - resources are deleted together with ISVCs
+	return nil
+}
+
+func (r *KserveRawRouteReconciler) createDesiredResource(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) (*v1.Route, error) {
+	var err error
+	enableAuth := false
+	if enableAuth, err = strconv.ParseBool(isvc.Labels[constants.LabelEnableAuthODH]); err != nil {
+		enableAuth = false
+	}
+	createRoute := false
+	if val, ok := isvc.Labels[constants.KserveNetworkVisibility]; ok && val == constants.LabelEnableRoute {
+		createRoute = true
+	}
+	if !createRoute {
+		log.Info("InferenceService does not have label '" + constants.KserveNetworkVisibility + "' annotation" +
+			" set to '" + constants.LabelEnableRoute + "'. Skipping route creation")
+		return nil, nil
+	}
+
+	// Fetch the service with the label "serving.kserve.io/inferenceservice=isvc.Name" in the isvc namespace
+	serviceList := &corev1.ServiceList{}
+	labelSelector := client.MatchingLabels{"serving.kserve.io/inferenceservice": isvc.Name}
+	err = r.client.List(ctx, serviceList, client.InNamespace(isvc.Namespace), labelSelector)
+	if err != nil || len(serviceList.Items) == 0 {
+		log.Error(err, "Failed to fetch service for InferenceService", "InferenceService", isvc.Name)
+		return nil, err
+	}
+	targetServiceName := serviceList.Items[0].Name
```

Comment:
You can have more than one if the `InferenceService` has transformers, predictors,etc... In that case, I don't think you can blindly use the first one.